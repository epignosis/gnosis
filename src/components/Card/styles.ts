import { css, Theme, SerializedStyles } from "@emotion/react";
import { hexToRGBA } from "@theme/default/colors";

export const cardContainer = ({ card }: Theme): SerializedStyles => css`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
  background-color: ${card.background};
  box-shadow: 0px 0px 10px #00000026;
  border-radius: 5px;
  overflow: hidden;
  transition-duration: 0.2s;
  transition-property: background-color, color, box-shadow 0.1s ease-in-out;

  :hover {
    box-shadow: 0px 4px 20px ${card.shadow};
    transition: all 0.2s ease;
  }
`;

export const courseHeaderContainer = css`
  position: relative;
  height: 100%;
  width: 100%;
`;

export const thumbnailContainer = css`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

export const hoverContainer = (
  { card }: Theme,
  { transparent }: { transparent?: boolean },
): SerializedStyles => {
  return css`
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    background-color: ${transparent ? "transparent" : card.hover.background};

    .hover-wrapper {
      height: 100%;
      width: 100%;
    }
  `;
};

export const cardBody = css`
  padding: 0.75rem 1rem 1.5rem;
`;

export const overlayContainer = ({ card }: Theme): SerializedStyles => css`
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  background-color: ${hexToRGBA(card.overlay.background, 0.6)};
  z-index: 2;
`;

export const drawerContainer = ({ card }: Theme): SerializedStyles => css`
  position: absolute;
  top: 0;
  right: 0;
  height: 100%;
  background-color: ${card.background};
  z-index: 3;

  .content-wrapper {
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding: 1rem;

    header {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-bottom: 1rem;

      /* TODO: Remove when our Button link variant is ready */
      button {
        background-color: transparent;
        padding: 0;
        border: 0;
      }
    }

    footer {
      margin-top: 0.5rem;
    }
  }
`;
