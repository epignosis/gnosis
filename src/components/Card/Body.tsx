import React, { FC } from "react";
import { cardBody } from "./styles";

export type BodyProps = FC;

const Body: BodyProps = ({ children }) => <section css={cardBody}>{children}</section>;

export default Body;
