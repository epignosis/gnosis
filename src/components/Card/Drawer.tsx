import React, { FC, ReactNode } from "react";
import { m, AnimatePresence, Variants, Transition } from "framer-motion";
import { useClickAway } from "@umijs/hooks";
import { CloseSVG } from "../../icons/";
import { drawerContainer } from "./styles";

export type DrawerProps = FC<{
  isOpen: boolean;
  onClose: () => void;
  title?: ReactNode;
  footer?: ReactNode;
}>;

const transition: Transition = {
  duration: 0.25,
  ease: "easeInOut",
};

const drawerContainerVariants: Variants = {
  open: {
    width: "100%",
    transition: {
      ...transition,
      when: "beforeChildren",
    },
    transitionEnd: {
      overflowY: "auto",
    },
  },
  closed: {
    width: "0%",
    overflowY: "hidden",
    transition: {
      ...transition,
      when: "afterChildren",
    },
  },
};

const drawerContentVariants: Variants = {
  open: {
    opacity: 1,
  },
  closed: {
    opacity: 0,
  },
};

const Drawer: DrawerProps = ({ isOpen, onClose, title, footer, children }) => {
  const containerRef = useClickAway((): void => {
    if (isOpen) {
      onClose();
    }
  });

  return (
    <AnimatePresence>
      {isOpen && (
        <m.section
          css={drawerContainer}
          ref={containerRef}
          initial="closed"
          animate={isOpen && "open"}
          exit="closed"
          variants={drawerContainerVariants}
        >
          <m.div className="content-wrapper" variants={drawerContentVariants}>
            <div>
              <header>
                <div>{title}</div>
                <button onClick={onClose}>
                  <CloseSVG height={16} />
                </button>
              </header>
              <p>{children}</p>
            </div>
            {footer && <footer>{footer}</footer>}
          </m.div>
        </m.section>
      )}
    </AnimatePresence>
  );
};

export default Drawer;
