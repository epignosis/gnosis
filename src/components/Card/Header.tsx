import React, { FC } from "react";
import AspectRatio from "../AspectRatio/AspectRatio";
import { courseHeaderContainer } from "./styles";

export type CardHeaderProps = FC<{
  ratio?: [number, number];
}>;

const CardHeader: CardHeaderProps = ({ ratio = [4, 3], children }) => {
  return (
    <AspectRatio ratio={ratio}>
      <section css={courseHeaderContainer}>{children}</section>
    </AspectRatio>
  );
};

export default CardHeader;
