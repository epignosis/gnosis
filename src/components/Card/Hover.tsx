import React, { FC } from "react";
import { m, Variants } from "framer-motion";
import { SerializedStyles } from "@emotion/react";
import { hoverContainer } from "./styles";

export type HoverProps = FC<{
  transparent?: boolean;
}>;

const hoverContainerVariants: Variants = {
  rest: {
    opacity: 0,
  },
  hover: {
    opacity: 1,
  },
};

const hoverContent: Variants = {
  rest: {
    display: "none",
  },
  hover: {
    display: "block",
  },
};

const Hover: HoverProps = ({ transparent = false, children }) => (
  <m.div
    transition={{
      duration: 0.15,
      ease: "easeIn",
    }}
    variants={hoverContainerVariants}
    css={(theme): SerializedStyles => hoverContainer(theme, { transparent })}
  >
    <m.div className="hover-wrapper" variants={hoverContent}>
      {children}
    </m.div>
  </m.div>
);

export default Hover;
