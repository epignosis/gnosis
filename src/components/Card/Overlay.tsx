import React, { FC } from "react";
import { overlayContainer } from "./styles";

export type OverlayProps = FC;

const Overlay: OverlayProps = ({ children }) => <div css={overlayContainer}>{children}</div>;

export default Overlay;
