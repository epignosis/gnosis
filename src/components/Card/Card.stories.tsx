import React, { useReducer } from "react";
import { Story } from "@storybook/react";
import { Button, Heading, Text, ProgressBar } from "../../";
import { PlayUnitSVG } from "../../icons/";
import Card from "./Card";

export default {
  title: "components/Card",
  component: Card,
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 300 }}>
        <Story />
      </div>
    ),
  ],
};

const Template: Story<{ isOverlayShow: boolean; isDrawerOpen: boolean }> = ({
  isOverlayShow,
  isDrawerOpen,
}) => {
  const [isOpen, toggle] = useReducer((state) => !state, isDrawerOpen);

  return (
    <Card>
      {isOverlayShow && <Card.Overlay>Hello Overlay</Card.Overlay>}
      <Card.Drawer
        isOpen={isOpen}
        title={
          <>
            <Text fontSize="md" weight="700">
              Course code:{" "}
            </Text>
            SAMP1
          </>
        }
        onClose={toggle}
        footer={<Button block>Read more</Button>}
      >
        Lorem ipsum, dolor sit amet consectetur adipisicing elit. Quibusdam labore ratione error
        adipisci dolorem corporis voluptatibus quia nihil quaerat ipsa tempora, in neque facere
        doloremque aut tempore ullam molestias necessitatibus!
      </Card.Drawer>

      <Card.Header>
        <Card.Thumbnail
          src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
          alt="An image"
        />
        {!isOverlayShow && (
          <Card.Hover>
            <div
              style={{
                height: "100%",
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
              }}
            >
              <Button size="lg" onClick={toggle} rounded>
                <PlayUnitSVG height={48} />
              </Button>
            </div>
          </Card.Hover>
        )}
      </Card.Header>
      <Card.Body>
        <div
          style={{
            display: "flex",
            justifyContent: "space-between",
            alignItems: "center",
            marginBottom: "0.5rem",
          }}
        >
          <a>Something</a>
          <span>Details</span>
        </div>
        <Heading as="h3" size="md">
          Introduction to TalentLMS
        </Heading>
      </Card.Body>
      <ProgressBar percent={55} rounded={false} />
    </Card>
  );
};

export const KitchenSink = Template.bind({});

KitchenSink.args = {
  isOverlayShow: false,
  isDrawerOpen: false,
};
