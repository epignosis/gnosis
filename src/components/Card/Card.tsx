import React, { FC } from "react";
import { AnimatePresence, domAnimation, LazyMotion, m, MotionProps } from "framer-motion";
import { cardContainer } from "./styles";
import Header, { CardHeaderProps } from "./Header";
import Thumbnail, { ThumbnailProps } from "./Thumbnail";
import Hover, { HoverProps } from "./Hover";
import Body, { BodyProps } from "./Body";
import Overlay, { OverlayProps } from "./Overlay";
import Drawer, { DrawerProps } from "./Drawer";

export type CardProps = FC<React.HTMLAttributes<HTMLElement> & MotionProps>;

type CardCompoundProps = {
  Header: CardHeaderProps;
  Thumbnail: ThumbnailProps;
  Hover: HoverProps;
  Body: BodyProps;
  Overlay: OverlayProps;
  Drawer: DrawerProps;
};

const Card: CardProps & CardCompoundProps = ({ children, ...rest }) => (
  <LazyMotion features={domAnimation}>
    <AnimatePresence>
      <m.article initial="rest" whileHover="hover" css={cardContainer} data-testid="card" {...rest}>
        {children}
      </m.article>
    </AnimatePresence>
  </LazyMotion>
);

Card.Header = Header;
Card.Thumbnail = Thumbnail;
Card.Hover = Hover;
Card.Body = Body;
Card.Overlay = Overlay;
Card.Drawer = Drawer;

export default Card;
