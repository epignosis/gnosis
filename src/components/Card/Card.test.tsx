import React from "react";
import faker from "faker";
import userEvent from "@testing-library/user-event";
import Card from "./Card";
import { screen, render } from "@test-utils/render";

describe("<Card/>", () => {
  const alt = faker.lorem.word();
  const src = faker.image.imageUrl();
  const bodyTxt = faker.lorem.paragraph();
  const hoverTxt = faker.lorem.word();
  const overlayTxt = faker.lorem.word();
  const titleTxt = faker.lorem.word();
  const footerTxt = faker.lorem.words();
  const mockFn = jest.fn();

  it("renders correctly", () => {
    render(
      <Card>
        <Card.Header>
          <Card.Thumbnail src={src} alt={alt} />
        </Card.Header>
        <Card.Body>{bodyTxt}</Card.Body>
      </Card>,
    );

    const img = screen.getByAltText(alt);
    const body = screen.getByText(bodyTxt);

    expect(img).toHaveAttribute("src", src);
    expect(body).toHaveTextContent(bodyTxt);
  });

  it("has hover action", () => {
    render(
      <Card>
        <Card.Header>
          <Card.Thumbnail src={src} alt={alt} />
          <Card.Hover>{hoverTxt}</Card.Hover>
        </Card.Header>
      </Card>,
    );

    const hover = screen.getByText(hoverTxt);

    expect(hover).not.toBeVisible();

    // userEvent.hover(hover);

    // expect(screen.getByText(hoverTxt)).toBeVisible();
  });

  it("has overlay", () => {
    render(
      <Card>
        <Card.Overlay>{overlayTxt}</Card.Overlay>
        <Card.Header>
          <Card.Thumbnail src={src} alt={alt} />
        </Card.Header>
      </Card>,
    );

    const overlay = screen.getByText(overlayTxt);

    expect(overlay).toHaveTextContent(overlayTxt);
  });

  it("has drawer", () => {
    render(
      <Card>
        <Card.Drawer isOpen title={titleTxt} onClose={mockFn} footer={footerTxt}>
          {bodyTxt}
        </Card.Drawer>
      </Card>,
    );

    const title = screen.getByText(titleTxt);
    const body = screen.getByText(bodyTxt);
    const footer = screen.getByText(footerTxt);
    const closeBtn = screen.getByRole("button");

    expect(title).toHaveTextContent(titleTxt);
    expect(body).toHaveTextContent(bodyTxt);
    expect(footer).toHaveTextContent(footerTxt);

    userEvent.click(closeBtn);

    expect(mockFn).toHaveBeenCalledTimes(1);

    userEvent.click(document.body);

    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Card>
        <Card.Header>
          <Card.Thumbnail
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
            alt="An image"
          />
        </Card.Header>
        <Card.Body>
          <a>Something</a>
          <span>Details</span>
        </Card.Body>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with hover", () => {
    const { container } = render(
      <Card>
        <Card.Header>
          <Card.Thumbnail
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
            alt="An image"
          />
          <Card.Hover>Hover text</Card.Hover>
        </Card.Header>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with transparent hover", () => {
    const { container } = render(
      <Card>
        <Card.Header>
          <Card.Thumbnail
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
            alt="An image"
          />
          <Card.Hover transparent>Hover text</Card.Hover>
        </Card.Header>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with overlay", () => {
    const { container } = render(
      <Card>
        <Card.Overlay>Hello Overlay</Card.Overlay>

        <Card.Header>
          <Card.Thumbnail
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
            alt="An image"
          />
        </Card.Header>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with drawer", () => {
    const { container } = render(
      <Card>
        <Card.Drawer isOpen title="Drawer title" onClose={mockFn} footer="Drawer footer">
          Lorem ipsum, dolor sit amet consectetur adipisicing elit. Quibusdam labore ratione error
          adipisci dolorem corporis voluptatibus quia nihil quaerat ipsa tempora, in neque facere
          doloremque aut tempore ullam molestias necessitatibus!
        </Card.Drawer>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });
});
