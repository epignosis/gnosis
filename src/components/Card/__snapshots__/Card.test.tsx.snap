// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> matches snapshot 1`] = `
<div>
  <article
    class="main-card css-u1k06h-cardContainer-cardContainer"
    data-testid="card"
    id="my-card"
  >
    <div
      class="css-14dee1b-aspectRatioContainer-aspectRatioContainer"
    >
      <div
        class="aspect-ratio-wrapper"
      >
        <section
          class="css-pcvco1-courseHeaderContainer"
        >
          <img
            alt="An image"
            class="css-13ekei5-thumbnailContainer"
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
          />
        </section>
      </div>
    </div>
    <section
      class="css-1m2q5o0-cardBody"
    >
      <a>
        Something
      </a>
      <span>
        Details
      </span>
    </section>
  </article>
</div>
`;

exports[`<Card/> matches snapshot with drawer 1`] = `
<div>
  <article
    class="css-u1k06h-cardContainer-cardContainer"
    data-testid="card"
  >
    <section
      class="css-1siuu8c-drawerContainer-drawerContainer"
      style="width: 0%; overflow-y: hidden;"
    >
      <div
        class="content-wrapper"
        style="opacity: 0;"
      >
        <div>
          <header>
            <div>
              Drawer title
            </div>
            <button>
              <svg-file-stub
                height="16"
              />
            </button>
          </header>
          <p>
            Lorem ipsum, dolor sit amet consectetur adipisicing elit. Quibusdam labore ratione error adipisci dolorem corporis voluptatibus quia nihil quaerat ipsa tempora, in neque facere doloremque aut tempore ullam molestias necessitatibus!
          </p>
        </div>
        <footer>
          Drawer footer
        </footer>
      </div>
    </section>
  </article>
</div>
`;

exports[`<Card/> matches snapshot with hover 1`] = `
<div>
  <article
    class="css-u1k06h-cardContainer-cardContainer"
    data-testid="card"
  >
    <div
      class="css-14dee1b-aspectRatioContainer-aspectRatioContainer"
    >
      <div
        class="aspect-ratio-wrapper"
      >
        <section
          class="css-pcvco1-courseHeaderContainer"
        >
          <img
            alt="An image"
            class="css-13ekei5-thumbnailContainer"
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
          />
          <div
            class="css-eo7ul9-hoverContainer-hoverContainer"
            data-testid="card-hover"
            style="opacity: 0;"
          >
            <div
              class="hover-wrapper"
              style="display: none;"
            >
              Hover text
            </div>
          </div>
        </section>
      </div>
    </div>
  </article>
</div>
`;

exports[`<Card/> matches snapshot with overlay 1`] = `
<div>
  <article
    class="css-u1k06h-cardContainer-cardContainer"
    data-testid="card"
  >
    <div
      class="css-12e19c2-overlayContainer-overlayContainer"
    >
      Hello Overlay
    </div>
    <div
      class="css-14dee1b-aspectRatioContainer-aspectRatioContainer"
    >
      <div
        class="aspect-ratio-wrapper"
      >
        <section
          class="css-pcvco1-courseHeaderContainer"
        >
          <img
            alt="An image"
            class="css-13ekei5-thumbnailContainer"
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
          />
        </section>
      </div>
    </div>
  </article>
</div>
`;

exports[`<Card/> matches snapshot with transparent hover 1`] = `
<div>
  <article
    class="css-u1k06h-cardContainer-cardContainer"
    data-testid="card"
  >
    <div
      class="css-14dee1b-aspectRatioContainer-aspectRatioContainer"
    >
      <div
        class="aspect-ratio-wrapper"
      >
        <section
          class="css-pcvco1-courseHeaderContainer"
        >
          <img
            alt="An image"
            class="css-13ekei5-thumbnailContainer"
            src="https://talentlms-prod-frontend-static.s3.us-east-1.amazonaws.com/images/default-course.png"
          />
          <div
            class="css-qtcg2f-hoverContainer-hoverContainer"
            data-testid="card-hover"
            style="opacity: 0;"
          >
            <div
              class="hover-wrapper"
              style="display: none;"
            >
              Hover text
            </div>
          </div>
        </section>
      </div>
    </div>
  </article>
</div>
`;
